source/Mlos.Python/mlos/Optimizers/OptimizerBase.py

```diff
@@ -87,39 +90,53 @@ def optimum(self, optimum_definition: OptimumDefinition = OptimumDefinition.BEST
         """
         assert optimum_definition in OptimumDefinition
 
-        features_df, objectives_df = self.get_all_observations()
-        if not len(features_df.index):
+        parameters_df, objectives_df, _ = self.get_all_observations()
+
+        if not len(parameters_df.index):
             raise ValueError("Can't compute optimum before registering any observations.")
+        have_context = (context is not None) or (self.optimization_problem.context_space is not None)
+        if have_context and optimum_definition != OptimumDefinition.BEST_SPECULATIVE_WITHIN_CONTEXT:
+            raise ValueError(f"{optimum_definition} not supported if context is provided.")

```

Comment:
I think this was probably already discussed so sorry if I missed/forgot it.
This will throw an error when the optimization problem includes a context space as well, not just if the `optimum()` caller passes context, right?
Is the idea that we aren't going to do anything about the question of "optimum() over all contexts" for now (or ever)?

Reply:
yes exactly. Sorry I discussed that in the 1:1 with Adam. We concluded that doing max makes little sense (i.e. it would always find the configuration that works best on the easiest context, like the biggest machine or smallest workload).
And adding some more robust thing is left for a future PR.